cores/arduino/stm32/rtc.c

```diff
@@ -0,0 +1,426 @@
+#else
+    UNUSED(subSeconds);
+#endif
+    if(format == PM) {
+      RTC_AlarmStructure.AlarmTime.TimeFormat = RTC_HOURFORMAT12_PM;
+    } else {
+      RTC_AlarmStructure.AlarmTime.TimeFormat = RTC_HOURFORMAT12_AM;
+    }
+    RTC_AlarmStructure.AlarmTime.DayLightSaving = RTC_DAYLIGHTSAVING_NONE;
+    RTC_AlarmStructure.AlarmTime.StoreOperation = RTC_STOREOPERATION_RESET;
+    RTC_AlarmStructure.AlarmDateWeekDay = date;
+    RTC_AlarmStructure.AlarmDateWeekDaySel = RTC_ALARMDATEWEEKDAYSEL_DATE;
+    RTC_AlarmStructure.AlarmMask = RTC_ALARMMASK_NONE;
+#else
+    UNUSED(subSeconds);
+    UNUSED(format);
+    UNUSED(date);
+#endif // !defined(STM32F1xx)
+
+    /* Set RTC_Alarm */
+    HAL_RTC_SetAlarm_IT(&RtcHandle, &RTC_AlarmStructure, RTC_FORMAT_BIN);
+  }
+}
+
+/**
+  * @brief Disable RTC alarm
+  * @param None
+  * @retval None
+  */
+void RTC_StopAlarm(void)
+{
+  /* Clear RTC Alarm Flag */
+  __HAL_RTC_ALARM_CLEAR_FLAG(&RtcHandle, RTC_FLAG_ALRAF);
+
+  /* Disable the Alarm A interrupt */
+  HAL_RTC_DeactivateAlarm(&RtcHandle, RTC_ALARM_A);
+}
+
+/**
+  * @brief Get RTC alarm
+  * @param date: 1-31 (day of the month)
+  * @param hours: 0-12 or 0-23 depends on the hours mode.
+  * @param minutes: 0-59
+  * @param seconds: 0-59
+  * @param subSeconds: 0-999
+  * @param hours format: AM or PM
+  * @retval None
+  */
+void RTC_GetAlarm(uint8_t *date, uint8_t *hours, uint8_t *minutes, uint8_t *seconds, uint32_t *subSeconds, hourAM_PM_t *format)
+{
+  RTC_AlarmTypeDef RTC_AlarmStructure;
+
+  if((date != NULL) && (hours != NULL) && (minutes != NULL) && (seconds != NULL) && (subSeconds != NULL) && (format != NULL)) {
+    HAL_RTC_GetAlarm(&RtcHandle, &RTC_AlarmStructure, RTC_ALARM_A, RTC_FORMAT_BIN);
+
+    *seconds = RTC_AlarmStructure.AlarmTime.Seconds;
+    *minutes = RTC_AlarmStructure.AlarmTime.Minutes;
+    *hours = RTC_AlarmStructure.AlarmTime.Hours;
+#if !defined(STM32F1xx)
+    *date = RTC_AlarmStructure.AlarmDateWeekDay;
+    if(RTC_AlarmStructure.AlarmTime.TimeFormat == RTC_HOURFORMAT12_PM) {
+      *format = PM;
+    } else {
+      *format = AM;
+    }
+#if !defined(STM32F2xx) && !defined(STM32L1xx)
+    *subSeconds = RTC_AlarmStructure.AlarmTime.SubSeconds;
+#elif defined(STM32L100xBA) || defined (STM32L151xBA) || defined (STM32L152xBA) || defined(STM32L100xC) || defined (STM32L151xC) || defined (STM32L152xC) || defined (STM32L162xC) || defined(STM32L151xCA) || defined (STM32L151xD) || defined (STM32L152xCA) || defined (STM32L152xD) || defined (STM32L162xCA) || defined (STM32L162xD) || defined(STM32L151xE) || defined(STM32L151xDX) || defined (STM32L152xE) || defined (STM32L152xDX) || defined (STM32L162xE) || defined (STM32L162xDX)
+    *subSeconds = RTC_AlarmStructure.AlarmTime.SubSeconds;
+#endif
+#endif // !defined(STM32F1xx)
+  }
+}
+
+/**
+  * @brief Attach alarm callback.
+  * @param func: pointer to the callback
+  * @retval None
+  */
+void attachAlarmCallback(void (*func)(void))
```

Comment:
rather than 
void (*func)(void)
it could be useful from application to offer a callback of type
void (*func)(void* data)

Reply:
What information would you like to pass in parameter? rtc handler? other?

Reply:
This is up to application .  It could the time when alarm was set, or a pointer to the led to be turned on. The application provides a pointer and you simply pass it back to application. This is useful for callback mecanisms.  It would be particularly useful if setting several alarms.

Reply:
It is not the application that choice this parameter. It is the driver which can return a parameter.
If you don't say me the parameter to return, I think return a void* is useless.
Or I must change the prototype of attachAlarmCallback to take account another parameter with the data pointer (also useless).

Reply:
sorry if unclear, the user data is passed from application to driver, stored in driver and then passed back from driver to application. The driver doesn't know which informaition it contains.

This can be useful for coding generic callback where you don't need to store application data as globals or in context.
https://arduino.stackexchange.com/questions/23063/user-callback-functions
https://stackoverflow.com/questions/14270846/passing-a-parameter-in-a-callback-in-c

In the case of RTC alarm, this can be useful if application can set 2 different alarms in future, then application could pass the programmed time, or a different LED or whatever it needs

Reply:
It is clearer thank you. And done.